internal/core/app_status_report.go

```diff
@@ -0,0 +1,221 @@
+
+	"github.com/hashicorp/go-argmapper"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl/v2"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	sdk "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+	"github.com/hashicorp/waypoint/internal/config"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+)
+
+// Builds a status report on the given deployment
+func (a *App) StatusReport(
+	ctx context.Context,
+	deployTarget *pb.Deployment,
+	releaseTarget *pb.Release,
+) (*pb.StatusReport, *sdk.StatusReport, error) {
+	// Query the artifact
+	var artifact *pb.PushedArtifact
+	if deployTarget != nil && deployTarget.Preload != nil &&
+		deployTarget.Preload.Artifact != nil {
+		artifact = deployTarget.Preload.Artifact
+	}
+	if artifact == nil && deployTarget != nil {
+		a.logger.Debug("querying artifact", "artifact_id", deployTarget.ArtifactId)
+		resp, err := a.client.GetPushedArtifact(ctx, &pb.GetPushedArtifactRequest{
+			Ref: &pb.Ref_Operation{
+				Target: &pb.Ref_Operation_Id{
+					Id: deployTarget.ArtifactId,
+				},
+			},
+		})
+		if err != nil {
+			a.logger.Error("error querying artifact",
+				"artifact_id", deployTarget.ArtifactId,
+				"error", err)
+			return nil, nil, err
+		}
+
+		artifact = resp
+	}
+
+	var evalCtx hcl.EvalContext
+	evalCtx.Variables = map[string]cty.Value{}
+	if err := evalCtxTemplateProto(&evalCtx, "artifact", artifact); err != nil {
+		a.logger.Warn("failed to prepare template variables, will not be available",
+			"err", err)
+	}
+
+	c, err := a.createStatusReporter(ctx, &evalCtx)
+	if status.Code(err) == codes.Unimplemented {
+		c = nil
+		err = nil
+	}
+	if err != nil {
+		a.logger.Error("error creating component in platform", "error", err)
+		return nil, nil, err
+	}
+	defer c.Close()
+
+	a.logger.Debug("starting status report operation")
+	statusReporter, ok := c.Value.(component.Status)
+
+	if !ok || statusReporter.StatusFunc() == nil {
+		a.logger.Debug("component is not a Status or has no StatusFunc()")
+		return nil, nil, nil
+	}
+
+	result, msg, err := a.doOperation(ctx, a.logger.Named("statusreport"), &statusReportOperation{
+		Component:     c,
+		DeployTarget:  deployTarget,
+		ReleaseTarget: releaseTarget,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var status *sdk.StatusReport
+	if result != nil {
+		status = result.(*sdk.StatusReport)
+	}
+
+	return msg.(*pb.StatusReport), status, nil
+}
+
+func (a *App) createStatusReporter(
+	ctx context.Context,
+	hclCtx *hcl.EvalContext,
+) (*Component, error) {
+	log := a.logger
+
+	// Load variables from deploy
+	hclCtx = hclCtx.NewChild()
+	if _, err := a.deployEvalContext(ctx, hclCtx); err != nil {
+		return nil, err
+	}
+
+	log.Debug("initializing status report plugin")
+	// Potential bug here with k8s apply plugin
+	// Works with docker and k8s ok
+	c, err := componentCreatorMap[component.PlatformType].Create(ctx, a, hclCtx)
+	if err == nil {
+		// We have a status reporter configured, use that.
+		return c, nil
```

Comment:
nit: I think it's usually preferred to have the "happy path" as the last return

https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88
Also mentioned by Dave Cheney https://dave.cheney.net/2020/02/23/the-zen-of-go